controllers/argocd/statefulset.go

```diff
@@ -441,13 +441,56 @@ func getArgoControllerContainerEnv(cr *argoprojv1a1.ArgoCD) []corev1.EnvVar {
 	return env
 }
 
-func (r *ReconcileArgoCD) reconcileApplicationControllerStatefulSet(cr *argoprojv1a1.ArgoCD, useTLSForRedis bool) error {
+func (r *ReconcileArgoCD) getArgoControllerReplicaCount(cr *argoprojv1a1.ArgoCD) int32 {
 	var replicas int32 = common.ArgocdApplicationControllerDefaultReplicas
 
-	if cr.Spec.Controller.Sharding.Replicas != 0 && cr.Spec.Controller.Sharding.Enabled {
+	if cr.Spec.Controller.Sharding.EnableDynamicScaling {
+		if cr.Spec.Controller.Sharding.MinShards < 1 {
+			log.Info("Minimum number of shards cannot be less than 1. Setting default value to 1")
+			cr.Spec.Controller.Sharding.MinShards = 1
+		}
+
+		if cr.Spec.Controller.Sharding.MaxShards < cr.Spec.Controller.Sharding.MinShards {
```

Comment:
Should this check be done as part of validation (admission controller) when creating the CR ? Changing customer provided `spec` should be avoided.

Reply:
I would agree, we should not be writing back to the `spec`

we can always work with local variables rather than changing spec fields though

Reply:
@ishitasequeira @anandf 
hmmm is a validation webhook for this overkill? I think we can achieve this kind of min/max validation just through CRD openapi schema validation


```
// +kubebuilder:validation:Maximum
specifies the maximum numeric value that this field can have.

// +kubebuilder:validation:Minimum
specifies the minimum numeric value that this field can have. Negative numbers are supported.
```
https://book.kubebuilder.io/reference/markers/crd-validation.html

Reply:
@jaideepr97 kubebuilder could surely be used, but we also want to have validation where we check that the `maxReplicas > minReplicas`. This check is available as a beta feature in k8s v1.25 but not in earlier versions. Thus the use of validation webhook.

https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules

Reply:
By using kubebuilder annotations, we might be able to support these validations in `OCP 4.10` and `OCP 4.11` which use `1.23` and `1.24`versions of kubernetes respectively.

The problem with having validation logic in controller is that the user is not notified about it, until he finds a problem and sets on to debug a problem by going through the controller logs. If some values in the CR is invalid, user should be notified about it during the CR creation time, so that the user gets a chance to fix the CR and re-create it. It would also avoid silent misconfigurations that can become a pain to debug.

@jaideepr97 Can we have this validation webhook code (even if its a trivial check) and slowly move all validations done in controller to this webhook in subsequent PRs (if we decide to go for it) ?

Reply:
@jaideepr97 following our internal conversations over [slack](https://redhat-internal.slack.com/archives/C01RQH8KQ87/p1684877099000859) around how we could prevent customers from having mis-configurations, we could utilize the validation webhook for addressing this issues. 

Also, I agree to the point by @anandf about notifying customers during CR creation time rather than just adding logs around it and having customers go through logs to understand the problem.

Reply:
On further thought, I think I like the idea of moving all validation into the validating webhook
it does seem like a cleaner approach and would help clean up the operator code as well, so I am in favor of it

I would like @iam-veeramalla 's  thoughts on this as well though

Reply:
> moving all validation into the validating webhook

I would disagree that this is something we should do. The validation should happen at both places, similar to how you'd have validation in a frontend/backend scenario. You validate in the frontend (here: webhook) to let the user know early-on that something is wrong, but you validate in the backend (here: controller) as well to ensure you never process invalid data.

We could make functions for the validation logic, to avoid redundancy in code and to prevent validation logic potentially drifting apart.